ydb/core/kqp/opt/physical/effects/kqp_opt_phy_upsert_index.cpp

```diff
@@ -328,30 +330,167 @@ TExprBase MakeUpsertIndexRows(TKqpPhyUpsertIndexMode mode, const TDqPhyPrecomput
         .Done();
 }
 
-TMaybe<TCondenseInputResult> CheckUniqueConstraint(const TExprBase& inputRows, const THashSet<TStringBuf> inputColumns, bool checkOnlyGivenColumns,
-    const TKikimrTableDescription& table, const TSecondaryIndexes& indexes, TPositionHandle pos, TExprContext& ctx)
+TMaybe<TCondenseInputResult> CheckUniqueConstraint(const TExprBase& inputRows, const THashSet<TStringBuf> inputColumns,
+    const TKikimrTableDescription& table, const TSecondaryIndexes& indexes,
+    TPositionHandle pos, TExprContext& ctx)
 {
     auto condenseResult = CondenseInput(inputRows, ctx);
     if (!condenseResult) {
         return {};
     }
 
+    // In case of absent on of index columns in the UPSERT(or UPDATE ON) values
+    // we need to get actual value from main table to perform lookup in to the index table
+    THashSet<std::string_view> missedKeyInput;
+
     // Check uniq constraint for indexes which will be updated by input data.
     // but skip main table pk columns - handle case where we have a complex index is a tuple contains pk
     const auto& mainPk = table.Metadata->KeyColumnNames;
     THashSet<TString> usedIndexes;
+    bool hasUniqIndex = false;
     for (const auto& [_, indexDesc] : indexes) {
+        hasUniqIndex |= (indexDesc->Type == TIndexDescription::EType::GlobalSyncUnique);
         for (const auto& indexKeyCol : indexDesc->KeyColumns) {
-            if (inputColumns.contains(indexKeyCol) &&
-                (std::find(mainPk.begin(), mainPk.end(), indexKeyCol) == mainPk.end()))
-            {
-                usedIndexes.insert(indexDesc->Name);
-                break;
+            if (inputColumns.contains(indexKeyCol)) {
+                if (!usedIndexes.contains(indexDesc->Name) &&
```

Comment:
we can omit `!usedIndexes.contains(indexDesc->Name)` check, the code will work the same